crates/codegen/parser/runtime/src/cursor.rs

```diff
@@ -394,3 +408,51 @@ impl Cursor {
         false
     }
 }
+
+/// A [`Cursor`] that also keeps track of the names of the nodes it visits.
+pub struct CursorWithNames {
```

Comment:
not sure I understand why do we need a new `CursorWithNames` cursor here, over returning an iterator object from the same `Cursor` type, similar to how `HashMap` can return different iterators via `.keys()` / `.values()` / etc...

Reply:
This is the typical iterator adapter pattern; In `std` these are not implemented with an iterator object (object, as in, involving dynamic dispatch) nor an existential type/impl Trait (i.e. `iter::from_fn`) but using regular wrapper/adapter structs: 

https://doc.rust-lang.org/stable/src/std/collections/hash/map.rs.html#1492-1494

This impl allows us to unwrap into the inner cursor, if desired (dropping the names), which wouldn't be possible with an anonymous type. Even though we're not using `size_hint` and few other iterator traits right now for the `Cursor`, we might be in the future.

Using the anonymous `iter::from_fn` comes at the cost, since it can't implement any other iterator traits due to the dynamic nature of a user-given closure-as-iterator, which would create asymmetry with the wrapped iterator - it should retain the same traits/characteristics; after all, this is just a convenience that yield the node together with its name.

Reply:
I see. Thanks for explaining!